### PR TITLE
[functionapp] - fix devops-build command

### PR DESCRIPTION
**Pull request by Hazhzeng**
This change contain a fix for the recently released `az functionapp devops-build create` command.

The following document contains a summary of fixes:

### Major Flow Change
1. Pipeline is now created based on Azure Devops repository (previously depends on Azure Devops project).
    - Discussion in my team concludes that even in a single project, we may have multiple repositories and multiple functionapps. We should allow customer to bind each functionapp with different repository.

### Major Fixes
1. Previously, when generating an Azure Devops pipeline build yaml file (a.k.a. azure-pipelines.yml), the jinja template failed to load in [azure-functions-devops-build library](https://github.com/Azure/azure-functions-devops-build)
    - This is due to the `azure_functions_devops_build/yaml/templates/build.jinja` template is not included in PyPi package manifest (see azure-functions-devops-build changes [here](https://github.com/Azure/azure-functions-devops-build/commit/deb592de2c4ac4450c568eea80387bd9bc24e379))
2. Previously, the CLI fails creating multiple pipelines from a local functionapp
    - This is due to the collision of "devopsbuild" remote in local git context. (see new functions **process_local_repository** and **process_remote_repository** in `azure_devops_build_iteractive.py`, and removal of **process_repository**)
3. Previously, the release fails due to the missing artifact information from incomplete build in Azure Devops release pipeline.
    - This is due to the CLI will keep polling the build artifacts once a build is triggered (and usually, the build artifact will get published before the build is complete), we should check the build.result instead. (see changes in **process_release** in `azure_devops_build_iteractive.py`)

### Minor Fixes
1. Change all `exit(1)` function in `azure_devops_build_iteractive.py` to `raise CLIError(message)`.
2. Add `git` detection. If user does not have git install, error out.
3. Add `git credential manager` detection. If user does not have credential manager install, we guide them to Azure Devops page and enable alternative credential settings.
4. Add Azure Devops repository dirty detection. If the remote Azure Devops repository is dirty, we should perform a force push (after acquiring user's consent)
5. Allow CLI to rollback remote in local git context when fail to create a pipeline (see [azure-functions-devops-build](https://github.com/Azure/azure-functions-devops-build/pull/22/files))
6. Fix an issue where CLI cannot detect existing Azure Devops **service connection**, **build definition** and **release definition**.
7. Show an error message when user does not have sufficient permission to create a service connection for Azure Devops pipeline.

### Development Environment
1. Add two main files `appservice\azure_devops_build_iteractive.py` and `appservice\azure_devops_build_provider.py` in solution file, so they are included in Visual Studio.

### Scenario Test
1. Add End-to-End test `appservice\tests\latest\test_devops_build_commands.py`
    - **test_devops_build_command** checks if the command can successfully deploy from local context to Azure functionapp.
    - **test_devops_build_mismatch_runtime** checks the error handling of mismatched local runtime and remote functionapp runtime
    - **test_devops_build_mismatch_functionapp** checks the error handling when the name of a functionapp does not exist in Azure.
    - **test_devops_build_mismatch_organization, test_devops_build_mismatch_project** check the error hanlding when a user does not have access to an Azure Devops organization or project.
2. Add End-to-End test sample functionapp folder `appservice\tests\latest\sample_dotnet_function`
    - This folder contains a basic sample dotnet functionapp created by [Functionapp Core Tool](https://github.com/Azure/azure-functions-core-tools), to support the scenario tests.

cc @anirudhgarg @Francisco-Gamino @ColbyTresness @dariagrigoriu @ankitkumarr @maiqbal11
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by Hazhzeng** [comment]
Hi @yugangw-msft, the change fails flake8 static checks. Since the `az functionapp devops-build create` command is currently broken in production. This fix needs to push out as soon as possible.

Can we start the Code Review without fixing the linting issue? Please let me know, thanks.

cc @anirudhgarg

**Comment by tjprescott** [comment]
@Hazhzeng you need to fix the static checks. You can run `azdev style` locally to identify them quickly and iterate until they are fixed, but allowing this to be merged in that state will cause all builds to fail.

**Comment by ahmedelnably** [comment]
@Hazhzeng can you please also change the help text to:
az functionapp dev-ops-build - Please capitalize Azure DevOps so it would read "Azure Function specific integration with Azure DevOps"
az functionapp devops-build create - Capitalize DevOps, and remove the word "build" so it would read "Create an Azure DevOps pipeline for a function app"

**Comment by Hazhzeng** [comment]
Thanks @tjprescott, I fixed the CI issues in travis, please let me know if this is ready.
Also, thanks @ahmedelnably for picking out the capitalization issues in command help text.

cc: @ColbyTresness

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/azure_devops_build_iteractive.py [hunk 1]
```diff
@@ -75,21 +82,28 @@ class (the rest are helpers)
...
         self.process_service_endpoint()
         self.process_extensions()
 
+        # Start build process and release artifacts to azure functions app
+        self.process_build_and_release_definition_name()
         self.process_build()
         self.process_release()
 
+        # Advise user to reuse the pipeline build and release by pushing to remote
+        print()
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/azure_devops_build_iteractive.py]: Suggest use `logger.warning()`. `print` is generally discouraged for lack of lots of advantage of using python logging, also printed text goes to `stdout`, and for consistency, should only contain data, not status/error.
For information you don't like users to see by default, use `logger.info()` or `logger.debug()`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py [hunk 2]
```diff
@@ -401,8 +400,7 @@ def load_arguments(self, _):
...
         c.argument('organization_name', help="Name of the Azure DevOps organization that you want to use", required=False)
         c.argument('project_name', help="Name of the Azure DevOps project that you want to use", required=False)
+        c.argument('repository_name', help="Name of the Azure Devops repository that you want to use", required=False)
         c.argument('overwrite_yaml', help="If you have an existing yaml, should it be overwritten?", arg_type=get_three_state_flag(return_label=True), required=False)
+        c.argument('allow_force_push', help="If Azure Devops repository is not clean, should it overwrite remote content?", arg_type=get_three_state_flag(return_label=True), required=False)
         c.argument('use_local_settings', help="Use your local settings in your functionapp settings?", arg_type=get_three_state_flag(return_label=True), required=False)
-        c.argument('local_git', help="If you want have a local git repository in this folder, what should we do? Adding a remote will preserve your local repository and use a remote to reference the build repository. "
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py]: This is a breaking change, unless this guy never works. If yes, please add an entry in the history.rst to call this out

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/azure_devops_build_iteractive.py (hunk 1 above)
Comment (Hazhzeng) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/azure_devops_build_iteractive.py]: Thanks @yugangw-msft, I've changed all print statement to logger statements.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py (hunk 2 above)
Comment (Hazhzeng) [reply src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py]: @yugangw-msft, I updated HISTORY.rst in appservice to reflect the breaking change. Thanks for your time.

**Comment by Hazhzeng** [comment]
The issues have been fixed and I rebase my code on the latest 'dev' branch.
Please let me know when it is good to go :)

**Pull request merged**